File: graphedu/api/services/system/dict.py
"""字典管理路由 (Controller层)

职责：
1. 接收HTTP请求，使用DTO进行参数验证
2. 调用Service层处理业务逻辑
3. 响应VO对象给前端
"""

from fastapi import APIRouter, Depends, Path, Query
from redis.asyncio import Redis as AsyncRedis
from sqlalchemy.ext.asyncio import AsyncSession

from graphedu.common.models.constants import SystemConstants
from graphedu.common.models.dto.systemv2.dict import (
    DictDataCreateDTO,
    DictDataQueryDTO,
    DictDataUpdateDTO,
    DictTypeCreateDTO,
    DictTypeQueryDTO,
    DictTypeUpdateDTO,
)
from graphedu.common.models.vo import (
    DictDataDetailVO,
    DictDataListVO,
    DictDataSimpleVO,
    DictTypeDetailVO,
    DictTypeListVO,
)
from graphedu.common.models.vo.base import Empty, PageResponse, ResponseType, ResponseUtil
from graphedu.common.resource.deps import get_db, get_redis
from graphedu.security.aspect.interface_auth import CheckUserInterfacePermit
from graphedu.security.aspect.log_annotation import SystemLog
from graphedu.security.auth import SecurityService
from graphedu.services.system.dict import DictDataService, DictTypeService

# 创建路由
dict_controller = APIRouter(prefix="/system/dict")


# ============== 字典类型管理 ==============


@dict_controller.get(
    "/type/list",
    dependencies=[Depends(CheckUserInterfacePermit("system:dict:list"))],
    response_model=ResponseType[PageResponse[DictTypeListVO]],
)
async def get_dict_type_list(dict_type_query: DictTypeQueryDTO = Query(), query_db: AsyncSession = Depends(get_db)):
    """获取字典类型列表"""
    dict_type_query_result = await DictTypeService.get_dict_type_list(query_db, dict_type_query)
    return ResponseUtil.success(data=dict_type_query_result)


@dict_controller.post(
    "/type", dependencies=[Depends(CheckUserInterfacePermit("system:dict:add"))], response_model=ResponseType[Empty]
)
@SystemLog(title="字典管理", business_type=SystemConstants.BusinessType.INSERT)
async def add_dict_type(
    dict_type_add: DictTypeCreateDTO,
    current_user=Depends(SecurityService.get_current_user),
    query_db: AsyncSession = Depends(get_db),
    redis_session: AsyncRedis = Depends(get_redis),
):
    """新增字典类型"""
    await DictTypeService.add_dict_type(query_db, redis_session, dict_type_add, current_user)
    return ResponseUtil.success()


@dict_controller.put(
    "/type", dependencies=[Depends(CheckUserInterfacePermit("system:dict:edit"))], response_model=ResponseType
)
@SystemLog(title="字典管理", business_type=SystemConstants.BusinessType.UPDATE)
async def update_dict_type(
    dict_type_update: DictTypeUpdateDTO,
    current_user=Depends(SecurityService.get_current_user),
    query_db: AsyncSession = Depends(get_db),
    redis_session: AsyncRedis = Depends(get_redis),
):
    """修改字典类型"""
    await DictTypeService.update_dict_type(query_db, redis_session, dict_type_update, current_user)
    return ResponseUtil.success()


@dict_controller.delete(
    "/type/{dict_ids}",
    dependencies=[Depends(CheckUserInterfacePermit("system:dict:remove"))],
    response_model=ResponseType[Empty],
)
@SystemLog(title="字典管理", business_type=SystemConstants.BusinessType.DELETE)
async def delete_dict_type(
    dict_ids: str = Path(..., pattern="^[0-9,]+$", description="字典类型ID，多个以逗号分隔"),
    query_db: AsyncSession = Depends(get_db),
    redis_session: AsyncRedis = Depends(get_redis),
):
    """删除字典类型"""
    delete_id_list = [int(dict_id) for dict_id in dict_ids.split(",") if dict_ids]

    await DictTypeService.delete_dict_type(query_db, redis_session, delete_id_list)
    return ResponseUtil.success()


@dict_controller.get(
    "/type/{dict_id}",
    dependencies=[Depends(CheckUserInterfacePermit("system:dict:query"))],
    response_model=ResponseType[DictTypeDetailVO],
)
async def get_dict_type_detail(dict_id: int = Path(), query_db: AsyncSession = Depends(get_db)):
    """获取字典类型详细信息"""
    dict_type_vo = await DictTypeService.get_dict_type_detail(query_db, dict_id)
    return ResponseUtil.success(data=dict_type_vo)


@dict_controller.delete(
    "/type/refreshCache",
    dependencies=[Depends(CheckUserInterfacePermit("system:dict:remove"))],
    response_model=ResponseType[Empty],
)
@SystemLog(title="字典管理", business_type=SystemConstants.BusinessType.CLEAN)
async def refresh_dict_cache(query_db: AsyncSession = Depends(get_db), redis_session: AsyncRedis = Depends(get_redis)):
    """刷新字典缓存"""
    await DictTypeService.refresh_cache(query_db, redis_session)
    return ResponseUtil.success()


@dict_controller.get(
    "/type/optionselect",
    response_model=ResponseType[list[DictTypeListVO]],
)
async def get_dict_type_options(query_db: AsyncSession = Depends(get_db)):
    """获取字典类型选项列表（用于下拉框等场景）"""
    dict_type_options = await DictTypeService.get_dict_type_options(query_db)
    return ResponseUtil.success(data=dict_type_options)


# @dict_controller.post(
#     '/type/export', response_model=ResponseType,
#     dependencies=[Depends(CheckUserInterfacePermit('system:dict:export'))]
# )
# @log(title='字典管理', business_type=5)
# async def export_dict_type(
#         request: Request,
#         dict_type_query: DictTypeQueryDTO = Body(),
#         query_db: AsyncSession = Depends(get_db)
# ):
#     """
#     导出字典类型
#     """
#     # TODO: 实现导出功能
#     return ResponseUtil.success(message='导出功能待实现')


# ============== 字典数据管理 ==============


@dict_controller.get(
    "/data/list",
    dependencies=[Depends(CheckUserInterfacePermit("system:dict:list"))],
    response_model=ResponseType[PageResponse[DictDataListVO]],
)
async def get_dict_data_list(dict_data_query: DictDataQueryDTO = Query(), query_db: AsyncSession = Depends(get_db)):
    """获取字典数据列表"""
    dict_data_query_result = await DictDataService.get_dict_data_list(query_db, dict_data_query)
    return ResponseUtil.success(data=dict_data_query_result)


@dict_controller.get("/data/type/{dict_type}", response_model=ResponseType[list[DictDataSimpleVO]])
async def get_dict_data_by_type(
    dict_type: str = Path(), query_db: AsyncSession = Depends(get_db), redis_session: AsyncRedis = Depends(get_redis)
):
    """根据字典类型查询字典数据（优先从缓存获取，缓存不存在则从数据库加载）"""
    dict_data_list = await DictDataService.get_dict_data_from_cache(query_db, redis_session, dict_type)
    return ResponseUtil.success(data=dict_data_list)


@dict_controller.post(
    "/data", dependencies=[Depends(CheckUserInterfacePermit("system:dict:add"))], response_model=ResponseType[Empty]
)
@SystemLog(title="字典数据", business_type=SystemConstants.BusinessType.INSERT)
async def add_dict_data(
    dict_data_add: DictDataCreateDTO,
    current_user=Depends(SecurityService.get_current_user),
    query_db: AsyncSession = Depends(get_db),
    redis_session: AsyncRedis = Depends(get_redis),
):
    """新增字典数据"""
    await DictDataService.add_dict_data(query_db, redis_session, dict_data_add, current_user)
    return ResponseUtil.success()


@dict_controller.put(
    "/data", dependencies=[Depends(CheckUserInterfacePermit("system:dict:edit"))], response_model=ResponseType[Empty]
)
@SystemLog(title="字典数据", business_type=SystemConstants.BusinessType.UPDATE)
async def update_dict_data(
    dict_data_update: DictDataUpdateDTO,
    current_user=Depends(SecurityService.get_current_user),
    query_db: AsyncSession = Depends(get_db),
    redis_session: AsyncRedis = Depends(get_redis),
):
    """修改字典数据"""
    await DictDataService.update_dict_data(query_db, redis_session, dict_data_update, current_user)
    return ResponseUtil.success()


@dict_controller.delete(
    "/data/{dict_code_ids}",
    dependencies=[Depends(CheckUserInterfacePermit("system:dict:remove"))],
    response_model=ResponseType[Empty],
)
@SystemLog(title="字典数据", business_type=SystemConstants.BusinessType.DELETE)
async def delete_dict_data(
    dict_code_ids: str = Path(..., pattern="^[0-9,]+$", description="字典数据ID，多个以逗号分隔"),
    query_db: AsyncSession = Depends(get_db),
    redis_session: AsyncRedis = Depends(get_redis),
):
    """删除字典数据"""
    dict_code_id_list = [int(code) for code in dict_code_ids.split(",") if code]
    await DictDataService.delete_dict_data(query_db, redis_session, dict_code_id_list)
    return ResponseUtil.success()


@dict_controller.get(
    "/data/{dict_code}",
    dependencies=[Depends(CheckUserInterfacePermit("system:dict:query"))],
    response_model=ResponseType[DictDataDetailVO],
)
async def get_dict_data_detail(dict_code: int, query_db: AsyncSession = Depends(get_db)):
    """获取字典数据详细信息"""
    dict_data_vo = await DictDataService.get_dict_data_detail(query_db, dict_code)
    return ResponseUtil.success(data=dict_data_vo)


# @dict_controller.post('/data/export', response_model=ResponseType,
#                       dependencies=[Depends(CheckUserInterfacePermit('system:dict:export'))])
# @log(title='字典数据', business_type=5)
# async def export_dict_data(
#         request: Request,
#         dict_data_query: DictDataQueryDTO = Depends(DictDataQueryDTO.as_form),
#         query_db: AsyncSession = Depends(get_db)
# ):
#     """
#     导出字典数据
#     """
#     # TODO: 实现导出功能
#     return ResponseUtil.success()

File: graphedu-ui/src/api/system/dict.ts
/**
 * 字典管理相关 API
 * 对应后端：graphedu/api/services/system/dict.py
 */
import request from '@/utils/request'
import type { ResponseType, PageResponse, Empty } from '@/types/api/common.ts'
import type {
  DictTypeQueryDTO,
  DictTypeCreateDTO,
  DictTypeUpdateDTO,
  DictTypeListVO,
  DictTypeDetailVO,
  DictDataQueryDTO,
  DictDataCreateDTO,
  DictDataUpdateDTO,
  DictDataListVO,
  DictDataDetailVO,
  DictDataSimpleVO,
} from '@/types/api/system/dict.ts'

// ============== 字典类型管理 ==============

/**
 * 获取字典类型列表（分页）
 * GET /system/dict/type/list
 */
export function getDictTypeList(query: DictTypeQueryDTO): Promise<ResponseType<PageResponse<DictTypeListVO>>> {
  return request({
    url: '/system/dict/type/list',
    method: 'get',
    params: query,
  })
}

/**
 * 新增字典类型
 * POST /system/dict/type
 */
export function addDictType(data: DictTypeCreateDTO): Promise<ResponseType<Empty>> {
  return request({
    url: '/system/dict/type',
    method: 'post',
    data: data,
  })
}

/**
 * 修改字典类型
 * PUT /system/dict/type
 */
export function updateDictType(data: DictTypeUpdateDTO): Promise<ResponseType<Empty>> {
  return request({
    url: '/system/dict/type',
    method: 'put',
    data: data,
  })
}

/**
 * 删除字典类型
 * DELETE /system/dict/type/{dict_ids}
 */
export function deleteDictType(dictIds: string): Promise<ResponseType<Empty>> {
  return request({
    url: `/system/dict/type/${dictIds}`,
    method: 'delete',
  })
}

/**
 * 获取字典类型详细信息
 * GET /system/dict/type/{dict_id}
 */
export function getDictTypeDetail(dictId: number): Promise<ResponseType<DictTypeDetailVO>> {
  return request({
    url: `/system/dict/type/${dictId}`,
    method: 'get',
  })
}

/**
 * 刷新字典缓存
 * DELETE /system/dict/type/refreshCache
 */
export function refreshDictCache(): Promise<ResponseType<Empty>> {
  return request({
    url: '/system/dict/type/refreshCache',
    method: 'delete',
  })
}

/**
 * 获取字典类型选项列表（用于下拉框等场景）
 * GET /system/dict/type/optionselect
 */
export function getDictTypeOptions(): Promise<ResponseType<DictTypeListVO[]>> {
  return request({
    url: '/system/dict/type/optionselect',
    method: 'get',
  })
}

// ============== 字典数据管理 ==============

/**
 * 获取字典数据列表（分页）
 * GET /system/dict/data/list
 */
export function getDictDataList(query: DictDataQueryDTO): Promise<ResponseType<PageResponse<DictDataListVO>>> {
  return request({
    url: '/system/dict/data/list',
    method: 'get',
    params: query,
  })
}

/**
 * 根据字典类型查询字典数据（优先从缓存获取）
 * GET /system/dict/data/type/{dict_type}
 */
export function getDictDataByType(dictType: string): Promise<ResponseType<DictDataSimpleVO[]>> {
  return request({
    url: `/system/dict/data/type/${dictType}`,
    method: 'get',
  })
}

/**
 * 新增字典数据
 * POST /system/dict/data
 */
export function addDictData(data: DictDataCreateDTO): Promise<ResponseType<Empty>> {
  return request({
    url: '/system/dict/data',
    method: 'post',
    data: data,
  })
}

/**
 * 修改字典数据
 * PUT /system/dict/data
 */
export function updateDictData(data: DictDataUpdateDTO): Promise<ResponseType<Empty>> {
  return request({
    url: '/system/dict/data',
    method: 'put',
    data: data,
  })
}

/**
 * 删除字典数据
 * DELETE /system/dict/data/{dict_code_ids}
 */
export function deleteDictData(dictCodeIds: string): Promise<ResponseType<Empty>> {
  return request({
    url: `/system/dict/data/${dictCodeIds}`,
    method: 'delete',
  })
}

/**
 * 获取字典数据详细信息
 * GET /system/dict/data/{dict_code}
 */
export function getDictDataDetail(dictCode: number): Promise<ResponseType<DictDataDetailVO>> {
  return request({
    url: `/system/dict/data/${dictCode}`,
    method: 'get',
  })
}
